include/stl2/detail/memory/uninitialized_copy.hpp

```diff
@@ -41,24 +41,39 @@ STL2_OPEN_NAMESPACE {
 		return {std::move(first), std::move(result)};
 	}
 
+	///////////////////////////////////////////////////////////////////////////
+	// uninitialized_copy [Extension]
+	//
 	template <InputRange Rng, __NoThrowForwardIterator O>
 	requires
-		Constructible<value_type_t<O>, reference_t<iterator_t<Rng>>> &&
-		__ReferenceTo<O, value_type_t<O>>
+		Constructible<value_type_t<O>, reference_t<iterator_t<Rng>>>
 	tagged_pair<tag::in(safe_iterator_t<Rng>), tag::out(O)>
 	uninitialized_copy(Rng&& rng, O result)
 	{
 		return __stl2::uninitialized_copy(
 			__stl2::begin(rng), __stl2::end(rng), result);
 	}
 
+	///////////////////////////////////////////////////////////////////////////
+	// uninitialized_copy [Extension]
+	//
+	template <InputRange IRng, class T>
+	requires
+		Constructible<T, reference_t<iterator_t<IRng>>>
+	tagged_pair<tag::in(safe_iterator_t<IRng>), tag::out(iterator_t<ext::span<T>>)>
+	uninitialized_copy(IRng&& irng, ext::span<T> orng)
+	{
+		STL2_EXPECT(__stl2::size(irng) == __stl2::size(orng));
```

Comment:
`size(InputRange)` isn't required to be valid: this algorithm is underconstrained.

Reply:
How about:
```c++
if constexpr(SizedRange<IRng>) {
  STL2_EXPECT(/*...*/);
}
```

Reply:
Not only this, but I think there's an overconstraint on the expectation (should be `size(irng) <= size(orng)` rather than `==`).

Reply:
My only concern with this is that it invites conflict between size types that are unsigned vs signed (e.g. vector and span).

Reply:
My only concern with this is that it invites conflict between size types that are unsigned vs signed (e.g. vector and span). I've been sitting on this for the past few hours, but haven't come up with a good solution yet.

Reply:
Use `distance` - which is always signed - instead of `size`?

Reply:
Great idea (and I'll be using `distance` from now on, forever), but I'm keen on following the discussion above in favour of this.